include/rmm/device_scalar.hpp

```diff
@@ -92,6 +147,47 @@ class device_scalar {
 
  private:
   rmm::device_buffer buff{sizeof(T)};
+
+  template<bool synchronize>
+  inline T _value(cudaStream_t stream) const {
+    T host_value{};
+    _memcpy<synchronize>(&host_value, buff.data(), sizeof(T), stream);
+    return host_value;
+  }
+
+  template<bool synchronize>
+  inline void _set_value(T host_value, cudaStream_t stream) {
+    _memcpy<synchronize>(buff.data(), &host_value, sizeof(T), stream);
+  }
+
+  template<bool synchronize>
+  inline void _memcpy(void *dst, const void *src, size_t count, cudaStream_t stream) const {
```

Comment:
@harrism @trxcllnt

I'm using `RMM_CHECK_CUDA`, but it seems strange to me. Is there a better way to go about it? The other options I considered include changing `RMM_CHECK_CUDA` to throw exceptions instead of return error codes, or making `_memcpy_` return `rmmError_t` (but that would obfuscate what caused the error, since we're doing both a copy and a sync).

Reply:
`RMM_CHECK_CUDA` isn't appropriate here as is. We'll need to introduce some new error checking macros that throw exceptions.

Reply:
Sounds good to me. Are you suggesting i should introduce a new macro (such as `RMM_RAISE_CUDA`) in this PR, or do you have plans to do that in the future?

Reply:
Don't introduce a new macro. Just use the explicit pattern you see everywhere else and I'll update it in the future.

Reply:
I've seen something along these lines in other places, but I'm not sure what information should be included in the exception, or how to go about formatting. Additional case statements could be added to support per-error exception types. I feel like this is part of a larger discussion, though, and shouldn't be a part of this PR.
```
#define RMM_RAISE_CUDA(call) rmmRaiseCuda(call, __FILE__, __LINE__)
inline void rmmRaiseCuda(cudaError_t error, const char * file, unsigned int line) {
  switch (error) {
  case cudaSuccess:
    break;
  default:
    throw new std::runtime_error("error");
  }
}
```